backend/internal/authn/authnservice.go

```diff
@@ -0,0 +1,297 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+// Package authn provides the implementation for handling authentication requests.
+package authn
+
+import (
+	"errors"
+	"time"
+
+	"github.com/asgardeo/thunder/internal/authn/constants"
+	authndto "github.com/asgardeo/thunder/internal/authn/dto"
+	"github.com/asgardeo/thunder/internal/authn/model"
+	"github.com/asgardeo/thunder/internal/flow"
+	flowconst "github.com/asgardeo/thunder/internal/flow/constants"
+	flowmodel "github.com/asgardeo/thunder/internal/flow/model"
+	"github.com/asgardeo/thunder/internal/oauth/jwt"
+	authzutils "github.com/asgardeo/thunder/internal/oauth/oauth2/authz/utils"
+	oauthmodel "github.com/asgardeo/thunder/internal/oauth/oauth2/model"
+	sessionmodel "github.com/asgardeo/thunder/internal/oauth/session/model"
+	sessionstore "github.com/asgardeo/thunder/internal/oauth/session/store"
+	sessionutils "github.com/asgardeo/thunder/internal/oauth/session/utils"
+	"github.com/asgardeo/thunder/internal/system/error/serviceerror"
+	"github.com/asgardeo/thunder/internal/system/log"
+	systemutils "github.com/asgardeo/thunder/internal/system/utils"
+)
+
+var loggerComponentName = "AuthenticationService"
+
+// AuthenticationServiceInterface defines the interface for authentication service.
+type AuthenticationServiceInterface interface {
+	Execute(sessionDataKey, flowID, actionID string,
+		inputs map[string]string) (*model.AuthNResponse, *serviceerror.ServiceError)
+}
+
+// AuthenticationService implements the AuthenticationServiceInterface.
+type AuthenticationService struct{}
+
+// NewAuthenticationService creates a new instance of AuthenticationService.
+func NewAuthenticationService() AuthenticationServiceInterface {
+	return &AuthenticationService{}
+}
+
+// Execute executes the authentication flow based on the provided data.
+func (s *AuthenticationService) Execute(sessionDataKey, flowID, actionID string,
+	inputs map[string]string) (*model.AuthNResponse, *serviceerror.ServiceError) {
+	logger := log.GetLogger().With(log.String(log.LoggerKeyComponentName, loggerComponentName))
+
+	if !isValidFlow(sessionDataKey, flowID) {
+		return nil, &constants.ErrorInvalidAuthFlow
+	}
+
+	var sessionData *sessionmodel.SessionData
+	appID := ""
+
+	if isNewFlow(sessionDataKey, flowID) {
+		var svcErr *serviceerror.ServiceError
+		sessionData, appID, svcErr = loadAndUpdateAuthZSession(sessionDataKey)
+		if svcErr != nil {
+			logger.Error("Failed to load session data", log.Any("serviceError", svcErr))
+			return nil, svcErr
+		}
+	}
+
+	flowStep, flowErr := flow.GetFlowService().Execute(appID, flowID, actionID,
```

Comment:
Here, isn't only the appID or the flowID used in a given time?

If thats the case, those two cases, mutually exclusive and we better to have two methods to reflect those two.